### PR TITLE
NMS-12948: SSLCertMonitor should include more details about the expir(ing|ed) certificate in reason codes

### PR DESCRIPTION
**Pull request by dschlenk**
### All Contributors

* [X] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [X] Have you made a comment in that issue which points back to this PR?
* [X] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [ ] If this is a new or updated feature, is there documentation for the new behavior?
* [ ] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

If you have made additions or changes to the documentation, or if you _need_ documentation for these code changes, please make sure a technical writer has looked it over.

Once the reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12948

### CODE REVIEW

File: features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java [hunk 1]
```diff
@@ -265,6 +288,12 @@ public PollStatus poll(final MonitoredService svc, final Map<String, Object> par
                 String reason = "IOException while polling address: " + ipAddr;
                 LOG.debug(reason, e);
                 serviceStatus = PollStatus.unavailable(reason);
+            } catch (CertificateEncodingException e) {
+                e.printStackTrace();
```
Comment (cgorantla) [inline features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java:292]: Can replace with LOG

File: features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java [hunk 2]
```diff
@@ -265,6 +288,12 @@ public PollStatus poll(final MonitoredService svc, final Map<String, Object> par
...
+            } catch (NoSuchAlgorithmException e) {
+                String reason = "NoSuchAlgorithException (SHA-1) while polling address: " + ipAddr;
+                LOG.debug(reason, e);
+                e.printStackTrace();
```
Comment (cgorantla) [inline features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java:296]: can omit `e.printStackTrace()`

File: features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java [hunk 3]
```diff
@@ -222,27 +226,46 @@ public PollStatus poll(final MonitoredService svc, final Map<String, Object> par
                 for (int i = 0; i < certs.length && !serviceStatus.isAvailable(); i++) {
                     if (certs[i] instanceof X509Certificate) {
                         X509Certificate certx = (X509Certificate) certs[i];
+                        String subject = certx.getSubjectDN().getName();
```
Comment (cgorantla) [inline features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java:229]: Are all of these guaranteed to be not null ?

**Review by cgorantla — COMMENTED** [review]
Thanks for the contribution @dschlenk.  Few minor things.

File: features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java (hunk 3 above)
Comment (dschlenk) [reply features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java:229]: The javadocs do not explicitly say whether or not they can be null, so I might as well add a check.

File: features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java [hunk 4]
```diff
@@ -222,27 +226,52 @@ public PollStatus poll(final MonitoredService svc, final Map<String, Object> par
...
                         LOG.debug("Checking validity against dates: [current: {}, valid: {}], NotBefore: {}, NotAfter: {}", calCurrent.getTime(), calValid.getTime(), certx.getNotBefore(), certx.getNotAfter());
                         calBefore.setTime(certx.getNotBefore());
                         calAfter.setTime(certx.getNotAfter());
                         if (calCurrent.before(calBefore)) {
-                            LOG.debug("Certificate is invalid, current time is before start time");
-                            serviceStatus = PollStatus.unavailable("Certificate is invalid, current time is before start time");
+                            reasonBuilder.append("Certificate with fingerprint '").append(fprint).append("' issued to ")
+                                    .append(subject).append(" by ").append(issuer)
+                                    .append(" is not yet valid. Current time is before start time. It is valid from ")
+                                    .append(certx.getNotBefore().toString()).append(" until ").append(certx.getNotAfter()).append(".");
```
Comment (cgorantla) [inline features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java:246]: Guess we should also check these elements to be non-null, `certx.getNotBefore().toString()` ,  `certx.getNotAfter()`

File: features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java [hunk 5]
```diff
@@ -222,27 +226,52 @@ public PollStatus poll(final MonitoredService svc, final Map<String, Object> par
                 for (int i = 0; i < certs.length && !serviceStatus.isAvailable(); i++) {
                     if (certs[i] instanceof X509Certificate) {
                         X509Certificate certx = (X509Certificate) certs[i];
+                        String subject = "";
+                        if (certx.getSubjectDN() != null && certx.getSubjectDN().getName() != null) {
+                            certx.getSubjectDN().getName();
```
[the remark this answers is not in the document]
Comment (dschlenk) [reply features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java]: yikes!

File: features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java (hunk 4 above)
Comment (dschlenk) [reply features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java:252]: OK, I'm now just marking the service unavailable if either of them are null prior to attempting to use them. I assume that's sensible?

Comment (cgorantla) [reply features/poller/monitors/core/src/main/java/org/opennms/netmgt/poller/monitors/SSLCertMonitor.java:252]: yeah, that seems fine to me.

**Review by cgorantla — APPROVED** [review]
Looks good to me.

**Pull request merged**